1_The_Physical/Level_2_Stateful_Guess_Point/Shared/Student/src/domain/student/core/student.ts

```diff
@@ -0,0 +1,137 @@
+  firstName?: FirstNameValidationError;
+  lastName?: LastNameValidationError;
+  email?: EmailValidationError;
+}
+
+export class Student implements AggregateRoot<StudentState> {
+  public readonly id: string;
+  public readonly state: StudentState;
+  public readonly events: DomainEvent[] = [];
+
+  private constructor(props: StudentState) {
+    this.id = uuid();
+    this.state = props;
+    this.events.push(new StudentCreated(this));
+  }
+
+  public static create(
+    props: StudentInputProps
+  ): Result<Student, StudentValidationError> {
+    const firstNameResult = FirstName.create(props.firstName);
+    const lastNameResult = LastName.create(props.lastName);
+    const emailResult = Result.combine(firstNameResult, lastNameResult).flatMap(
+      ([FirstName, LastName]) =>
+        Student.generateEmail(FirstName.value, LastName.value)
+    );
+
+    const errors: StudentValidationError = {};
+    if (firstNameResult.isFailure()) {
+      errors.firstName = firstNameResult.error;
+    }
+    if (lastNameResult.isFailure()) {
+      errors.lastName = lastNameResult.error;
+    }
+    if (emailResult.isFailure() && !(emailResult.error instanceof Array)) {
+      errors.email = emailResult.error;
+    }
+
+    if (Object.keys(errors).length > 0) {
+      return Result.failure(errors);
+    }
+
+    return Result.success(
+      new Student({
+        firstName: firstNameResult.value!,
+        lastName: lastNameResult.value!,
+        email: emailResult.value!,
+      })
+    );
+  }
+
+  public updateFirstName(
+    firstName: string
+  ): Result<Student, StudentValidationError> {
+    const oldFirstName = this.state.firstName.value;
+    const result = Student.create({
+      firstName,
+      lastName: this.state.lastName.value,
+    });
+
+    if (result.isSuccess()) {
+      const student = result.value!;
+      student.events.push(
+        new FirstNameUpdated(student, oldFirstName, firstName)
+      );
+    }
+
+    return result;
+  }
+
+  public updateLastName(
+    lastName: string
+  ): Result<Student, StudentValidationError> {
+    const oldLastName = this.state.lastName.value;
+    const result = Student.create({
+      firstName: this.state.firstName.value,
+      lastName,
+    });
+
+    if (result.isSuccess()) {
+      const student = result.value!;
+      student.events.push(new LastNameUpdated(student, oldLastName, lastName));
+    }
+
+    return result;
+  }
+
+  public get email() {
+    return this.state.email.value;
+  }
+
+  public get firstName() {
+    return this.state.firstName.value;
+  }
+
+  public get lastName() {
+    return this.state.lastName.value;
+  }
+
+  private static generateEmail(
+    firstName: string,
+    lastName: string
+  ): Result<Email, EmailValidationError> {
+    const firstNamePrefix = firstName.trim().slice(0, 2).toLowerCase();
+    const lastNamePrefix = lastName.trim().slice(0, 5).toLowerCase();
+
+    return Email.generate({ local: `${lastNamePrefix}${firstNamePrefix}` });
```

Comment:
Should you also enforce duplicate email validation?

Reply:
If I haven't missed anything, this use case was not requested, for now, we don't have to enforce duplicate email validation.